### PR TITLE
*: Support locate function

### PR DESCRIPTION
**Pull request by shenli**
See: https://dev.mysql.com/doc/refman/5.7/en/string-functions.html#function_locate

### CODE REVIEW

**Comment by ngaut** [comment]
LGTM

File: expression/substring_test.go [hunk 1]
```diff
@@ -128,3 +128,62 @@ func (s *testSubstringSuite) TestSubstringIndex(c *C) {
...
 }
+
+func (s *testSubstringSuite) TestLocate(c *C) {
+	tbl := []struct {
+		substr string
+		str    string
+		pos    int64
+		result int
+	}{
+		{"bar", "foobarbar", -1, 4},
```
Comment (qiuyesuifeng) [inline expression/substring_test.go]: How about null, like `locate('a', null)`, `locate(null, 'abc')`?

File: expression/substring.go [hunk 2]
```diff
@@ -202,3 +203,83 @@ func (f *FunctionSubstringIndex) Eval(ctx context.Context, args map[interface{}]
...
+}
+
+// Eval implements the Expression Eval interface.
+func (f *FunctionLocate) Eval(ctx context.Context, args map[interface{}]interface{}) (interface{}, error) {
+	// eval str
+	fs, err := f.Str.Eval(ctx, args)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	str, ok := fs.(string)
```
Comment (qiuyesuifeng) [inline expression/substring.go]: Seems MySQL support some none-string args like `SELECT LOCATE(1, 2)`? 
Any need to convert fs to string?

**Comment by shenli** [comment]
PTAL

**Comment by siddontang** [comment]
LGTM

**Pull request merged**